tla-io/src/main/scala/at/forsyte/apalache/io/quint/Quint.scala

```diff
@@ -371,29 +372,75 @@ class Quint(quintOutput: QuintOutput) {
-        val (fieldNames, quintVals) = quintArgs
-          .grouped(2)
-          .foldRight((List[String](), List[QuintEx]())) {
-            case (Seq(QuintStr(_, f), v), (fields, values)) => ((f :: fields), v :: values)
-            case (invalidArgs, _) =>
-              throw new QuintIRParseError(s"Invalid argument given to Rec ${invalidArgs}")
-          }
-        variadicApp { tlaVals =>
-          val fieldsAndArgs = fieldNames.zip(tlaVals)
-          tla.rowRec(rowVar, fieldsAndArgs: _*)
-        }(quintVals)
+    def record(rowVar: Option[String]): Converter = { quintArgs =>
+      // The quint Rec operator takes its field and value arguments
+      // via a variadic operator requiring field names passed as strings to
+      // be alternated with values. E.g.,
+      //
+      //    Rec("f1", 1, "f2", 2)
+      //
+      // So we first separate out the field names from the values, so we
+      // can make use of the existing combinator for variadic operators.
+      //
+      // Empty records are fine: those are the unit value.
+      val (fieldNames, quintVals) = quintArgs
+        .grouped(2)
+        .foldRight((List[String](), List[QuintEx]())) {
+          case (Seq(QuintStr(_, f), v), (fields, values)) => ((f :: fields), v :: values)
+          case (invalidArgs, _) =>
+            throw new QuintIRParseError(s"Invalid argument given to Rec ${invalidArgs}")
+        }
+      variadicApp { tlaVals =>
+        val fieldsAndArgs = fieldNames.zip(tlaVals)
+        tla.rowRec(rowVar, fieldsAndArgs: _*)
+      }(quintVals)
+    }
+
+    // Create a TLA variant
+    def variant(quintType: QuintType): Converter = {
+      val tlaType = typeConv.convert(quintType)
+      tlaType match {
+        case variantType: VariantT1 =>
+          binaryApp("variant",
+              (labelInstruction, expr) =>
+                // The builder requires a string literal, rather than a string expression
+                // so we have to build the converted TLA expression and extract its string value.
+                labelInstruction.flatMap {
+                  case ValEx(TlaStr(label)) => tla.variant(label, expr, variantType)
+                  case invalidLabel =>
+                    throw new QuintIRParseError(s"Invalid label found in application of variant ${invalidLabel}")
+                })
+        case _ => throw new QuintIRParseError(s"Invalid type inferred for application of variant ${quintType}")
+      }
+    }
 
+    // the quint builtin operator representing match expressions looks like
+    //
+    // matchVariant(expr, "F1", elim_1, ..., "Fn", elim_n)
+    //
+    // Where each `elim_i` is an operator applying to value wrapped in field `Fi` of a variant.
+    //
+    // This is converted into the following Apalache expression, using Apalache's variant operators:
+    //
+    // CASE VariantTag(expr) = "F1" -> elim_1(VariantGetUnsafe(expr))
+    //   [] ...
+    //   [] VariantTag(expr) = "Fn" -> elim_n(VariantGetUnsafe(expr))
```

Comment:
Should this be

```suggestion
    // CASE VariantTag(expr) = "F1" -> elim_1(VariantGetUnsafe("F1", expr))
    //   [] ...
    //   [] VariantTag(expr) = "Fn" -> elim_n(VariantGetUnsafe("Fn", expr))
```

?

Reply:
Yep! Thanks for the catch :D